Don't remove imports referenced by Javadoc comments.

tommythorsen opened the pull request:
This change adds some code to the function that removes unused imports,
to ensure that imports referenced by Javadoc comments (such as {@link
SomeClass}) are not stripped out.

The unit test for removing imports is extended to test this new
functionality.

tommythorsen commented:
This change is a work in progress for the time being. It won't actually work until version 3.6.8 of JavaParser is released, but you can have a look at the upcoming code.

matozoid commented:
JavaParser releases are weekly. Around next Sunday it will be available.

tommythorsen commented:
Thanks for the review! As you might have guessed, I'm not super familiar with JavaDoc, but I will add the test cases from your comment, and make sure those work. Can you think of any other cases we need to handle?

ctubbsii commented:
I ran on a large project and those were the only ones I saw, but there were too many to check everything. There might be others. I can test again after the others are fixed and the noise is reduced. The project I tested with was https://github.com/apache/accumulo if you want to try it yourself `mvn clean package -DskipTests`

tommythorsen commented:
@ctubbsii I have added some more code that 1) also catches @see statements and 2) tokenizes the text for each javadoc tag, so that we  can support more complex usage.

I added all your examples to the unit test file.

Could you retest on accumulo? I tried to test it, but I was unable to reproduce the original errors with `mvn clean package -DskipTests` in the accumulo project.

ctubbsii commented:
To reproduce, you also have to enable this feature, `-Dimpsort.removeUnused=true` or by editing the configuration in the POM, since it's not enabled by default, and you have to modify the version in the POM to 1.1.1-SNAPSHOT.

I ran again with this latest, and the change set was much smaller (only 6 files were incorrectly modified; 7 total imports removed incorrectly). However, the class of problems looks like it was sometimes caught, and sometimes not... which leads me to believe there's a logic problem in the stream/filtering code... and not necessarily a problem with the matching logic.

ctubbsii commented:
I found that I could simplify the visitor to only process comments when visiting "JavadocComment" types, and it caught all but one. However, you have to pass the node directly, not `node.getComment()` and then convert to JavadocComment.

tommythorsen commented:
> I found that I could simplify the visitor to only process comments when visiting "JavadocComment" types, and it caught all but one. However, you have to pass the node directly, not node.getComment() and then convert to JavadocComment.

Done.

I was able to reproduce the one missing import in accumulo. The reason seems to be that there is an empty line between the comment and the class declaration, which makes this an orphaned comment. These are apparently not visited with the regular visitor. I'm working on a fix.

tommythorsen commented:
I have fixed the remaining problem by iterating over CompilationUnit.getComments() instead of using the VoidVisitor. It now works for accumulo.

ctubbsii commented:
Interesting that the VoidVisitor pattern didn't work. Seems like a bug in the visitor traversal or the Javadoc parsing in upstream javaparser? I liked the visitor pattern... was thinking about putting the visitor in its own class to clean up the ImpSort.java a bit, but this new code works too. Will test it and apply shortly.
